File: Problems/Algorithms/108. Convert Sorted Array to Binary Search Tree/SortedArrayToBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode aux(int[] nums, int start, int end) {
        if (start > end) return null;
        
        int mid = start + (end - start) / 2;
        TreeNode root = new TreeNode(nums[mid]);
        
        TreeNode left = aux(nums, start, mid-1);
        TreeNode right = aux(nums, mid+1, end);
        root.left = left;
        root.right = right;
        
        return root;
    }
    
    public TreeNode sortedArrayToBST(int[] nums) {
        int start = 0;
        int end = nums.length - 1;
        return aux(nums, start, end);
    }
}

File: Problems/Algorithms/108. Convert Sorted Array to Binary Search Tree/sorted_array_to_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def aux(nums: List[int], start: int, end: int) -> Optional[TreeNode]:
            if start > end:
                return None

            mid = start + (end - start) // 2
            root = TreeNode(nums[mid])

            left, right = aux(nums, start, mid-1), aux(nums, mid+1, end)
            root.left, root.right = left, right

            return root

        return aux(nums, 0, len(nums)-1)

